cis194/week4/tien/homework4/src/Homework4.hs

```diff
@@ -30,3 +30,42 @@ data Tree a
   = Leaf
   | Node Integer (Tree a) a (Tree a)
   deriving (Show, Eq)
+
+checkDepth :: Eq a => Tree a -> Integer
+checkDepth Leaf = (0 :: Integer)
+checkDepth (Node _ Leaf _ Leaf) = 1
+checkDepth (Node _ Leaf _ right@(Node _ _ _ _)) = 1 + checkDepth right
+checkDepth (Node _ left@(Node _ _ _ _) _ Leaf) = 1 + checkDepth left
+checkDepth (Node _ left@(Node _ _ _ _) _ right@(Node _ _ _ _)) =
+  1 + checkDepth left + checkDepth right
+
+insert :: Eq t => t -> Tree t -> Tree t
+insert x Leaf = Node 0 Leaf x Leaf
+insert x (Node i left c Leaf) = Node i left c (Node (i + 1) Leaf x Leaf)
+insert x (Node i Leaf c right) = Node i (Node (i + 1) Leaf x Leaf) c right
+insert x (Node i left c right)
+  | (checkDepth left) > (checkDepth right) = Node i left c (insert x right)
+  | otherwise = Node i (insert x left) c right
+
+foldTree :: (Foldable t, Eq a) => t a -> Tree a
+foldTree = foldr insert Leaf
+
+-- Exercise 3.1 --
+xor :: [Bool] -> Bool
+xor = odd . foldr (\x acc -> acc + (bool 0 1 x)) (0 :: Integer)
+
+-- Exercise 3.2 --
+map' :: (a -> b) -> [a] -> [b]
+map' f = foldr (\x acc -> f x : acc) []
+
+-- Exercise 3.3 --
+myFoldl :: (a -> b -> a) -> a -> [b] -> a
+myFoldl f base xs = foldr (\x acc -> f acc x) base (reverse xs)
```

Comment:
Thought exercise here: reverse before folding is not efficient. Because it's 2 foldings on the same list. Is there a way to only fold it once? (hint: a recursive function is the same of a stack and stack operations (push and pop) are reversing the list of values when they are popped out comparing to when they are pushed in).  Is there a way to create a recursive function by folding out of a list then trigger it with `base`?